src/main/java/org/apache/sling/i18n/impl/JcrResourceBundleProvider.java

```diff
@@ -791,4 +733,102 @@ public String toString() {
+                this.resourceBundle = resourceBundle;
+                this.serviceRegistration = serviceRegistration;
+            }
+        }
+
+        ResourceBundleRegistry(BundleContext bundleContext) {
+            this.bundleContext = bundleContext;
+        }
+
+        JcrResourceBundle getRB(Key key) {
+            Entry entry = bundleServiceRegistrations.get(key);
+            return entry != null ? entry.resourceBundle : null;
+        }
+
+        Collection<JcrResourceBundle> getRBs() {
+            return bundleServiceRegistrations.values().stream().map(e -> e.resourceBundle).collect(Collectors.toList());
+        }
+
+        void updateRB(Key key, JcrResourceBundle resourceBundle) {
+            if (isClosed.get()) {
+                return;
+            }
+            ServiceRegistration<ResourceBundle> serviceReg = bundleContext.registerService(ResourceBundle.class, resourceBundle, serviceProps(key));
+            Entry oldEntry = bundleServiceRegistrations.put(key, new Entry(resourceBundle, serviceReg));
+            if (oldEntry != null) {
+                oldEntry.serviceRegistration.unregister();
+            }
+            log.debug("[ResourceBundleRegistry.updateRB] Registry updated - Nr of entries: {} - Keys: {}", bundleServiceRegistrations.size(), bundleServiceRegistrations.keySet());
+        }
+
+        private static Dictionary<String, Object> serviceProps(Key key) {
+            Dictionary<String, Object> serviceProps = new Hashtable<>();
+            if (key.baseName != null) {
+                serviceProps.put("baseName", key.baseName);
+            }
+            serviceProps.put("locale", key.locale.toString());
+            return serviceProps;
+        }
+
+        void unregisterRB(Key key) {
+            if (isClosed.get()) {
+                return;
+            }
+            Entry oldEntry = bundleServiceRegistrations.remove(key);
+            if (oldEntry!= null) {
+                oldEntry.serviceRegistration.unregister();
+            } else {
+                log.warn("[ResourceBundleRegistry.unregisterRB] Could not find resource bundle service for {}", key);
+            }
+        }
+
+        void clear() {
+            if (isClosed.get()) {
+                return;
+            }
+            clearInternal();
+        }
+
+        private void clearInternal() {
+            log.debug("[ResourceBundleRegistry.clearInternal] Before - Nr of Keys: {} - Keys: {}", bundleServiceRegistrations.size(), bundleServiceRegistrations.keySet());
+            List<ServiceRegistration<ResourceBundle>> regs = new ArrayList<>();
+            Iterator<java.util.Map.Entry<Key, Entry>> iterator = bundleServiceRegistrations.entrySet().iterator();
+            while(iterator.hasNext()) {
+                regs.add(iterator.next().getValue().serviceRegistration);
+                iterator.remove();
+            }
```

Comment:
An alternative approach could be
- make the field `bundleServiceRegistrations` an `AtomicReference<ConcurrentHashMap<Key, Entry>>`
- swap the map out with `AtomicReference#getAndSet()`
- take all the time you want to traverse the "old" map and unregister its services

That might further narrow down the possibility of any races between one thread removing the map entries and another one reading or adding/updating entries. Not sure if that would prevent any meaningful error scenarios.